### PR TITLE
fix no_std build for memory-db.

### PR DESCRIPTION
**Pull request by cheme**
Version 12.3 broke no_std compatibility.
This PR fixes that.
The PR also removes some warning (add 'dyn' info on types).

### CODE REVIEW

File: memory-db/src/lib.rs [hunk 1]
```diff
@@ -369,6 +372,33 @@ where
...
+	KF::Key: MallocSizeOf,
+{
+	fn size_of(&self, ops: &mut MallocSizeOfOps) -> usize {
+		use core::mem::size_of;
+		let mut n = self.data.capacity() * (size_of::<T>() + size_of::<H>() + size_of::<usize>());
+		for (k, v) in self.data.iter() {
+			n += k.size_of(ops);
+			n += v.size_of(ops);
+		}
+		n
```
Comment (bkchr) [inline memory-db/src/lib.rs]: Why these line breaks? (I think all should fit on one line)

File: memory-db/src/lib.rs [hunk 2]
```diff
@@ -369,6 +372,33 @@ where
...
+	H::Out: MallocSizeOf,
+	T: MallocSizeOf,
+	KF: KeyFunction<H>,
+	KF::Key: MallocSizeOf,
+{
+	fn size_of(&self, ops: &mut MallocSizeOfOps) -> usize {
+		use core::mem::size_of;
+		let mut n = self.data.capacity() * (size_of::<T>() + size_of::<H>() + size_of::<usize>());
+		for (k, v) in self.data.iter() {
+			n += k.size_of(ops);
```
Comment (bkchr) [inline memory-db/src/lib.rs]:
```suggestion
			n += k.size_of(ops) + v.size_of(ops);
```

File: memory-db/src/lib.rs [hunk 3]
```diff
@@ -369,6 +372,33 @@ where
...
+			n += v.size_of(ops);
```
Comment (bkchr) [inline memory-db/src/lib.rs]:
```suggestion
```

File: memory-db/src/lib.rs [hunk 4]
```diff
@@ -354,6 +354,9 @@ where
 	}
 }
 
+// not no_std implementation requires that hasmap core
+// got its implementation in parity-util-mem
```
Comment (dvdplm) [inline memory-db/src/lib.rs]: This comment is unclear to me, can you rephrase it? Is it "The `no_std` implementation requires that `Hashmap` has an impl of `MallocSizeOf` in `parity-util-mem`" maybe?

File: memory-db/src/lib.rs [hunk 5]
```diff
@@ -369,6 +372,30 @@ where
 	}
 }
 
+// This is temporary code and should be in
+// parity-util-mem, but hashmap_core will probably
+// get replaced at some point by hashbrown so
+// this is a pragmatic temporary solution.
```
Comment (dvdplm) [inline memory-db/src/lib.rs]: Can you open an issue so we don't lose track of this?

**Comment by dvdplm** [comment]
@cheme Will you tag&publish also?

**Comment by cheme** [comment]
yes and yank 12.3 (I also will update the parity pr with heapsize removal).
